pymc4/gp/cov.py

```diff
@@ -0,0 +1,157 @@
+        ----------
+        X1 : tensor, array-like
+            First point
+        X2 : tensor, array-like
+            Second point
+        """
+        return self._kernel.apply(X1, X2, **kwargs)
+
+    def __add__(self, cov2):
+        return CovarianceAdd(self, cov2)
+
+    def __mul__(self, cov2):
+        return CovarianceProd(self, cov2)
+
+
+class Combination(Covariance):
+    r"""Combination of two or more covariance functions
+
+    Parameters
+    ----------
+    cov1 : pm.Covariance
+        First covariance function.
+    cov2 : pm.Covariance
+        Second covariance function.
+    """
+
+    def __init__(self, cov1, cov2, **kwargs):
+        self.kernel1 = cov1._kernel
+        self.kernel2 = cov2._kernel
+        self.feature_ndims = self.kernel1.feature_ndims
+        if self.kernel1.feature_ndims != self.kernel2.feature_ndims:
+            raise ValueError("Cannot combine kernels with different feature_ndims")
+        super().__init__(self.kernel1.feature_ndims, **kwargs)
+
+
+class CovarianceAdd(Combination):
+    r"""Addition of two or more covariance functions.
+
+    Parameters
+    ----------
+    feature_ndims : int
+        number of rightmost dims to include in kernel computation
+    """
+
+    def _init_kernel(self, feature_ndims, **kwargs):
+        return self.kernel1 + self.kernel2
+
+
+class CovarianceProd(Combination):
+    r"""Product of two or more covariance functions.
+
+    Parameters
+    ----------
+    feature_ndims : int
+        number of rightmost dims to include in kernel computation
+    """
+
+    def _init_kernel(self, feature_ndims, **kwargs):
+        return self.kernel1 * self.kernel2
+
+
+class Stationary(Covariance):
+    r"""Base class for all Stationary Covariance functions"""
+
+    @property
+    def length_scale(self):
+        r"""Length scale of the covariance function"""
+        return self._length_scale
+
+
+class ExpQuad(Stationary):
+    r"""Exponentiated Quadratic Stationary Covariance Function.
+    Commonly known as the Radial Basis Kernel Function.
+
+    .. math::
+       k(x, x') = \sigma^2 \mathrm{exp}\left[ -\frac{(x - x')^2}{2 l^2} \right]
+
+    where :math:`\sigma` = ``amplitude``
+          :math:`l` = ``length_scale``
+
+    Parameters
+    ----------
+    amplitude : tensor, array-like
+        The :math:`\sigma` parameter of RBF kernel, amplitude > 0
+    length_scale : tensor, array-like
+        The :math:`l` parameter of the RBF kernel
+    feature_ndims : int, optional
+        number of rightmost dims to include in kernel computation
+    kwargs : optional
+        Other keyword arguments that tfp's ``ExponentiatedQuadratic`` kernel takes
+    """
+
+    def __init__(self, amplitude, length_scale, feature_ndims=1, **kwargs):
+        if not tf.reduce_all(amplitude > 0.0):
+            raise ValueError("`amplitude` parameter must not contains zero or negative entries.")
+        self._amplitude = amplitude
+        self._length_scale = length_scale
```

Comment:
It will be important to include ARD using the FeatureScaled kernel. Ideally this would be automated such that a scalar length scale implies ARD is not desired, while an array-valued parameter invokes ARD.

Reply:
I have added it in the base ``Covariance`` class as it is applicable to all the kernels... Should we keep the ``scale_diag`` parameter an optional keyword argument or required argument?